scorex-basics/src/main/scala/scorex/network/peer/PeerDatabaseImpl.scala

```diff
@@ -21,18 +22,30 @@ class PeerDatabaseImpl(settings: Settings, filename: Option[String]) extends Pee
   private lazy val ownNonce = settings.nodeNonce
 
   override def addOrUpdateKnownPeer(address: InetSocketAddress, peerInfo: PeerInfo): Unit = {
-    val updatedPeerInfo = Option(whitelistPersistence.get(address)).map { case dbPeerInfo =>
-      val nonceOpt = peerInfo.nonce.orElse(dbPeerInfo.nonce)
-      val nodeNameOpt = peerInfo.nodeName.orElse(dbPeerInfo.nodeName)
-      PeerInfo(peerInfo.lastSeen, nonceOpt, nodeNameOpt)
-    }.getOrElse(peerInfo)
-    whitelistPersistence.put(address, updatedPeerInfo)
-    database.commit()
+    if (!isBlacklisted(address)) {
+      val updatedPeerInfo = Option(whitelistPersistence.get(address)).map { case dbPeerInfo =>
+        val nonceOpt = peerInfo.nonce.orElse(dbPeerInfo.nonce)
+        val nodeNameOpt = peerInfo.nodeName.orElse(dbPeerInfo.nodeName)
+        PeerInfo(peerInfo.lastSeen, nonceOpt, nodeNameOpt)
+      }.getOrElse(peerInfo)
+      whitelistPersistence.put(address, updatedPeerInfo)
+      database.commit()
+    }
   }
 
   override def blacklistPeer(address: InetSocketAddress): Unit = {
-    whitelistPersistence.remove(address)
-    if (!isBlacklisted(address)) blacklist += address.getHostName -> System.currentTimeMillis()
+    if (!isBlacklisted(address)) {
+      blacklist += address.getHostName -> System.currentTimeMillis()
+      new util.ArrayList(whitelistPersistence.keyList())
```

Comment:
You don't need util.ArrayList here, just

      whitelistPersistence.keyList()
        .filter(_.getHostName == address.getHostName)
        .foreach(whitelistPersistence.remove(_))